src/crh/poseidon/mod.rs

```diff
@@ -31,191 +49,210 @@ pub trait PoseidonRoundParams<F: PrimeField>: Default + Clone {
-        // function, one is set to the padding constant and rest are 0. Always keep
-        // the 1st input as 0
-        let input = vec![
-            F::from(ZERO_CONST),
-            xl,
-            xr,
-            F::from(PADDING_CONST),
-            F::from(ZERO_CONST),
-            F::from(ZERO_CONST),
-        ];
-
-        // Never take the first output
-        self.permute(&input)[1]
+        Ok(state)
     }
 
-    pub fn hash_4(&self, inputs: [F; 4]) -> F {
-        // Only 4 inputs to the permutation are set to the input of this hash
-        // function, one is set to the padding constant and one is set to 0. Always
-        // keep the 1st input as 0
-        let input = vec![
-            F::from(ZERO_CONST),
-            inputs[0],
-            inputs[1],
-            inputs[2],
-            inputs[3],
-            F::from(PADDING_CONST),
-        ];
-
-        // Never take the first output
-        self.permute(&input)[1]
+    fn apply_linear_layer(state: &Vec<F>, mds: &Vec<Vec<F>>) -> Vec<F> {
+        let mut new_state: Vec<F> = Vec::new();
+        for i in 0..state.len() {
+            let mut sc = F::zero();
+            for j in 0..state.len() {
+                let mij = mds[i][j];
+                sc += mij * state[j];
+            }
+            new_state.push(sc);
+        }
+        new_state
     }
 }
 
-pub struct PoseidonCRH<F: PrimeField, P: PoseidonRoundParams<F>> {
-    field: PhantomData<F>,
-    params: PhantomData<P>,
-}
+impl<F: PrimeField, P: Rounds> FixedLengthCRH for CRH<F, P> {
+    const INPUT_SIZE_BITS: usize = 32 * 8 * P::WIDTH;
+    type Output = F;
+    type Parameters = PoseidonParameters<F>;
 
-impl<F: PrimeField, P: PoseidonRoundParams<F>> PoseidonCRH<F, P> {
-    pub fn create_mds<R: Rng>(_rng: &mut R) -> Vec<Vec<F>> {
-        let mds_matrix = Vec::new();
-        mds_matrix
+    // Not sure what's the purpose of this function of we are going to pass parameters
+    fn setup<R: Rng>(rng: &mut R) -> Result<Self::Parameters, Error> {
+        Ok(Self::Parameters::generate(rng))
     }
 
-    pub fn create_round_consts<R: Rng>(_rng: &mut R) -> Vec<F> {
-        let round_consts = Vec::new();
-        round_consts
+    fn evaluate(parameters: &Self::Parameters, input: &[u8]) -> Result<Self::Output, Error> {
+        let eval_time = start_timer!(|| "PoseidonCRH::Eval");
+
+        if (input.len() * 8) > Self::INPUT_SIZE_BITS {
+            panic!(
+                "incorrect input length {:?} for width {:?}",
+                input.len() / 32,
+                P::WIDTH,
+            );
+        }
+
+        // Not giving expected results
+        // let elts: Vec<F> = input.to_field_elements().unwrap_or(Vec::new());
+
+        let mut buffer = vec![0u8; Self::INPUT_SIZE_BITS / 8];
+
+        buffer.iter_mut().zip(input).for_each(|(b, l_b)| *b = *l_b);
+
+        let f_inputs: Vec<F> = buffer
+            .chunks(32)
+            .map(|x| F::from_le_bytes_mod_order(x))
```

Comment:
Note that this can cause collisions for BN254, as the scalar field is only 254 bits, so everything in the range [p + 1,..., 2^256] will have a counterpart in [0..p]

Reply:
Should we instead handle this as some parameter on the `ScalarField` itself?

Reply:
There's `F::PARAMETERS::MODULUS_BITs`, which you can extract the maximum number of bytes from.